Win10: Use `GetSystemTimePreciseAsFileTime` directly

ChrisDenton opened the pull request:
On Windows 10 we can use `GetSystemTimePreciseAsFileTime` directly instead of lazy loading it (with a fallback).

ChrisDenton committed:
Win10: Use GetSystemTimePreciseAsFileTime directly

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8050530639/job/21986333186) [(plain)](https://github.com/rust-lang/rust/commit/7e6dccc47d257727c8135ec0d933155d42463e0e/checks/21986333186/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=ChrisDenton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_fcd944ef-a204-4ae7-97e8-22481ab8f143
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=d4df0abdbf4583a55bf8128f28575a295484e06c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_fcd944ef-a204-4ae7-97e8-22481ab8f143
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_fcd944ef-a204-4ae7-97e8-22481ab8f143
GITHUB_TRIGGERING_ACTOR=ChrisDenton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121633/merge
GITHUB_WORKFLOW_SHA=d4df0abdbf4583a55bf8128f28575a295484e06c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Error: 
   0: tests failed

FAILED TEST: tests/pass/shims/time.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "i686-pc-windows-msvc" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/tests/pass/shims" "tests/pass/shims/time.rs" "-Zmiri-disable-isolation" "--edition" "2021"
error: pass test got exit status: 1, but expected 0


error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/shims/time.stderr` to the actual output
--- tests/pass/shims/time.stderr
+++ <stderr output>
+error: unsupported operation: can't call foreign function `GetSystemTimePreciseAsFileTime` on OS `windows`
+  --> RUSTLIB/std/src/sys/pal/PLATFORM/time.rs:LL:CC
+   |
+LL |             c::GetSystemTimePreciseAsFileTime(&mut t.t);
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `GetSystemTimePreciseAsFileTime` on OS `windows`
+   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
+   = note: BACKTRACE:
+   = note: inside `std::sys::pal::PLATFORM::time::SystemTime::now` at RUSTLIB/std/src/sys/pal/PLATFORM/time.rs:LL:CC
+   = note: inside `std::time::SystemTime::now` at RUSTLIB/std/src/time.rs:LL:CC
+   = note: inside `std::time::SystemTime::now` at RUSTLIB/std/src/time.rs:LL:CC
+note: inside `main`
+  --> $DIR/time.rs:LL:CC
+   |
+LL |     let now1 = SystemTime::now();
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to 1 previous error
+error: aborting due to 1 previous error
+


error: there were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: unsupported operation: can't call foreign function `GetSystemTimePreciseAsFileTime` on OS `windows`
full stderr:
full stderr:
error: unsupported operation: can't call foreign function `GetSystemTimePreciseAsFileTime` on OS `windows`
##[error]  --> /checkout/library/std/src/sys/pal/windows/time.rs:71:13
   |
LL |             c::GetSystemTimePreciseAsFileTime(&mut t.t);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `GetSystemTimePreciseAsFileTime` on OS `windows`
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::pal::windows::time::SystemTime::now` at /checkout/library/std/src/sys/pal/windows/time.rs:71:13: 71:56
   = note: inside `std::time::SystemTime::now` at /checkout/library/std/src/time.rs:513:20: 513:43
   = note: inside `std::time::SystemTime::now` at /checkout/library/std/src/time.rs:513:20: 513:43
note: inside `main`
  --> tests/pass/shims/time.rs:20:16
   |
LL |     let now1 = SystemTime::now();

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
---
    tests/pass/shims/time.rs

test result: FAIL. 1 failed; 323 passed; 8 ignored;

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compiletest-f5946aa5fa5ac232 --quiet` (exit status: 1)
Build completed unsuccessfully in 0:00:47
```

</details>

ChrisDenton committed:
Add shim for GetSystemTimePreciseAsFileTime

This is exactly the same as GetSystemTimeAsFileTime except that it promises maximum precision.

ChrisDenton commented:
This adds `GetSystemTimePreciseAsFileTime` to miri, which is identical to `GetSystemTimeAsFileTime` except that it promises to use the highest precision clock available. Since miri didn't artificially lower the precision in the first place, there is no difference between the two functions.

RalfJung commented:
Should Miri be limiting the precision of the other one, if that's what the official implementation does?

RalfJung commented on src/tools/miri/src/shims/time.rs:
```diff
@@ -111,11 +111,12 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
     fn GetSystemTimeAsFileTime(
         &mut self,
         LPFILETIME_op: &OpTy<'tcx, Provenance>,
+        link_name: &str,
```
That's a confusing argument name, maybe `shim_name` or so? I first through "link" refers to "symlink" or so.^^ Also please make this the first argument.

ChrisDenton commented:
> Should Miri be limiting the precision of the other one, if that's what the official implementation does?

🤷 Technically it's left unspecified in the Microsoft docs so it could just be the same clock and it's free to change between versions.

In practice it has similar limitations to `GetTickCount` as mentioned in [Windows Time](https://learn.microsoft.com/en-us/windows/win32/sysinfo/windows-time)

> GetTickCount and GetTickCount64 are limited to the resolution of the system timer, which is approximately 10 milliseconds to 16 milliseconds

This can also be adjusted with [timeBeginPeriod](https://learn.microsoft.com/en-gb/windows/win32/api/timeapi/nf-timeapi-timebeginperiod). But as far as I can tell miri doesn't currently emulate system "ticks" (which are used for all sorts of timers and scheduling) and it would perhaps be odd to implement it only for one function.

RalfJung commented:
r=me on the Miri part

ChrisDenton commented:
Squished commits.

Nilstrieb commented on library/std/src/sys/pal/windows/c/windows_sys.rs:
```diff
@@ -345,6 +345,10 @@ extern "system" {
     pub fn GetSystemTimeAsFileTime(lpsystemtimeasfiletime: *mut FILETIME) -> ();
 }
 #[link(name = "kernel32")]
+extern "system" {
```
why are these individual extern blocks?

ChrisDenton replied:
Because that's what are binding generator generates. The [README](https://github.com/rust-lang/rust/blob/52f8aec14c616387c5f793687f2d9026de6c78ca/library/std/src/sys/pal/windows/c/README.md) has a few details about how this file is made.

ChrisDenton replied:
Basically it's all made by the [windows-bindgen](https://crates.io/crates/windows-bindgen) crate.

Nilstrieb replied:
ah

Noratrieb reviewed (approved):
looks good (Chris ~~begged~~ asked me to review it)

Nilstrieb commented:
@bors r+

bors commented:
:pushpin: Commit eb40adbd4604e86a1b0bd3c89d71cec3d2a7d077 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb eb40adbd4604e86a1b0bd3c89d71cec3d2a7d077 -->
<!-- homu: {"type":"Approved","sha":"eb40adbd4604e86a1b0bd3c89d71cec3d2a7d077","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
